lib/SILOptimizer/Transforms/ARCCodeMotion.cpp

```diff
@@ -375,16 +377,50 @@ bool RetainCodeMotionContext::requireIteration() {
 
 void RetainCodeMotionContext::initializeCodeMotionDataFlow() {
   // Find all the RC roots in the function.
+  llvm::DenseMap<SILValue, Atomicity> AtomicityOfRoot;
+
   for (auto &BB : *F) {
     for (auto &II : BB) {
+      auto RCInst = dyn_cast<RefCountingInst>(&II);
+      if (!RCInst)
+        continue;
+      // Remember the atomicity of the root. If any
+      // of the RC operations on a the root are atomic,
+      // then the root is considered to be atomic.
+      Atomicity InstAtomicity = (RCInst->getAtomicity() == Atomicity::Atomic)
+                                    ? Atomicity::Atomic
+                                    : Atomicity::NonAtomic;
+      //auto Root = getRCRoot(&II);
+      auto Root = getRCRoot(RCInst->getOperand(0));
+
+      auto newRoot = AtomicityOfRoot.count(Root);
+      auto &RootAtomicity = AtomicityOfRoot[Root];
+
+      if (!newRoot)
```

Comment:
@eeckstein I don't think so. RCRootIndex is only populated when we see e.g. a retain instruction. But we collect information in AtomicityOfRoot for ALL RC instructions (including releases, etc) that operate on this root. We need information about ALL RC instructions, because if at least one of them is atomic, then all the new RC instructions we insert for this root as part of the transformation need to be atomic to be on the safe side.